From d64e5a07047a0e36d98949deed9895c7ae39bf84 Mon Sep 17 00:00:00 2001
From: Michael-CY Lee <michael-cy.lee@mediatek.com>
Date: Tue, 3 Jun 2025 16:35:03 +0800
Subject: [PATCH 220/236] mtk: hostapd: Deauth single-link or legacy STA when
 AP triggers Adv-TTLM

When the AP triggers Adv-TTLM that disables some links, STAs that only
connect with those links should be deauthentication.

Signed-off-by: Michael-CY Lee <michael-cy.lee@mediatek.com>
---
 src/ap/drv_callbacks.c |  7 ++++++-
 src/ap/hostapd.c       | 27 +++++++++++++++++++++++++++
 src/ap/hostapd.h       |  1 +
 3 files changed, 34 insertions(+), 1 deletion(-)

diff --git a/src/ap/drv_callbacks.c b/src/ap/drv_callbacks.c
index 5f367ef51..234068035 100644
--- a/src/ap/drv_callbacks.c
+++ b/src/ap/drv_callbacks.c
@@ -1527,9 +1527,14 @@ void hostapd_event_attlm(struct hostapd_data *hapd, struct attlm_event *attlm_ev
 			/* FIXME do Neg-TTLM teardown to prevent overlap
 			 * with Adv-TTLM
 			 */
-			for_each_mld_link(p_hapd, hapd)
+			for_each_mld_link(p_hapd, hapd) {
 				ap_for_each_sta(p_hapd, teardown_sta_ttlm,
 						NULL);
+
+				if (mld->new_attlm.disabled_links &
+				    BIT(p_hapd->mld_link_id))
+					hostapd_deauth_single_link_sta(p_hapd);
+			}
 			break;
 		case EVENT_ATTLM_SWITCH_TIME_EXPIRED:
 			mld_indicate_disabled = true;
diff --git a/src/ap/hostapd.c b/src/ap/hostapd.c
index 070edbb98..ae6cf3ec6 100644
--- a/src/ap/hostapd.c
+++ b/src/ap/hostapd.c
@@ -429,6 +429,33 @@ static int hostapd_broadcast_wep_set(struct hostapd_data *hapd)
 
 #ifdef CONFIG_IEEE80211BE
 #ifdef CONFIG_TESTING_OPTIONS
+void hostapd_deauth_single_link_sta(struct hostapd_data *hapd)
+{
+	struct sta_info *sta;
+	struct mld_info *info;
+	int num_links, i;
+	u16 reason = WLAN_REASON_DEAUTH_LEAVING;
+
+	wpa_printf(MSG_DEBUG, "Deauthenticate single link STA on link_id=%u\n",
+		   hapd->mld_link_id);
+	for (sta = hapd->sta_list; sta; sta = sta->next) {
+		info = &sta->mld_info;
+		num_links = 0;
+
+		if (ap_sta_is_mld(hapd, sta)) {
+			for (i = 0; i < MAX_NUM_MLD_LINKS && num_links < 2; i++) {
+				if (info->links[i].valid)
+					num_links++;
+			}
+		}
+
+		if (num_links <= 1) {
+			hostapd_drv_sta_deauth(hapd, sta->addr, reason);
+			ap_sta_deauthenticate(hapd, sta, reason);
+		}
+	}
+}
+
 
 static void hostapd_link_remove_timeout_handler(void *eloop_data,
 						void *user_ctx)
diff --git a/src/ap/hostapd.h b/src/ap/hostapd.h
index acc4386dc..a07d5b6cf 100644
--- a/src/ap/hostapd.h
+++ b/src/ap/hostapd.h
@@ -1078,6 +1078,7 @@ bool hostapd_mld_is_first_bss(struct hostapd_data *hapd);
 void hostapd_mld_interface_freed(struct hostapd_data *hapd);
 int hostapd_enable_mld(struct hostapd_data *hapd);
 int hostapd_disable_mld(struct hostapd_data *hapd);
+void hostapd_deauth_single_link_sta(struct hostapd_data *hapd);
 
 #define for_each_mld_link(partner, self) \
 	dl_list_for_each(partner, &self->mld->links, struct hostapd_data, link)
-- 
2.45.2

